### PR TITLE
discord bot module

### PR DESCRIPTION
**Pull request by TroyDota**
Pull request checklist:

- [ ] `CHANGELOG.md` was updated, if applicable

### CODE REVIEW

**Comment by TroyDota** [comment]
Oh and there is a minor config change.
`
[steam]
secret_key=

[discord]
client_id=
client_secret=
redirect_uri=https://<streamer_domain>/login/discord_auth
`

**Comment by TroyDota** [comment]
https://steamcommunity.com/dev/apikey 

https://discordapp.com/developers

File: pajbot/models/user.py [hunk 1]
```diff
@@ -113,6 +116,17 @@ def __init__(self, *args, **kwargs):
         UserDuelStats, uselist=False, cascade="all, delete-orphan", passive_deletes=True, back_populates="user"
     )
 
+    @hybrid_property
+    def offcd(self):
+        if self.last_pair:
+            return (self.last_pair + timedelta(days=1)) < utils.now()  
+        return True
+
+    def _setcd(self, db_session):
```
Comment (DatGuy1) [inline pajbot/models/user.py]: How about making it simulate a command instead of this

File: pajbot/models/user_connection.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    # Twitch user ID
+    twitch_id = Column(TEXT, primary_key=True, nullable=False)
+
+    # Discord user id 
+    discord_user_id = Column(TEXT, primary_key=True, nullable=False)
+    disord_username = Column(TEXT, primary_key=True, nullable=False)
+
+    # Steam user id 
+    steam_id = Column(TEXT, primary_key=True, nullable=False)
```
Comment (DatGuy1) [inline pajbot/models/user_connection.py]: Mention what type of ID in comment

File: pajbot/modules/__init__.py [hunk 3]
```diff
@@ -74,6 +74,7 @@
 from pajbot.modules.viewerpoints import MassPointsModule
 from pajbot.modules.warning import WarningModule
 from pajbot.modules.wolfram import WolframModule
+from pajbot.modules.discord_bot import DiscordModule
```
Comment (DatGuy1) [inline pajbot/modules/__init__.py]: Alphabetical

File: pajbot/oauth_client_edit.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
+from flask_oauthlib.client import OAuth
```
Comment (DatGuy1) [inline pajbot/oauth_client_edit.py]: Too lazy to go over this then I'll just say seemsgood

File: pajbot/web/routes/base/user.py [hunk 5]
```diff
@@ -1,11 +1,28 @@
+import logging
```
Comment (DatGuy1) [inline pajbot/web/routes/base/user.py]: Same as above

File: uwsgi_shared.ini [hunk 6]
```diff
@@ -12,5 +12,7 @@ die-on-term = true
 
 memory-report = true
 
+buffer-size=32768
```
Comment (DatGuy1) [inline uwsgi_shared.ini]: Is this needed

File: pajbot/apiwrappers/twitch/helix.py [hunk 7]
```diff
@@ -54,13 +54,13 @@ def _fetch_all_pages(page_fetch_fn, *args, **kwargs):
         """Fetch all pages using a function that returns a list of responses and a pagination cursor
```
Comment (DatGuy1) [inline pajbot/apiwrappers/twitch/helix.py]: Did you run the reformat

File: pajbot/managers/discord_bot.py [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+        await cmd.call(message)
+
+
+class DiscordBotManager(object):
+
+    def __init__(self, bot, redis):
+        self.bot = bot
+        self.client = CustomClient(self)
+
+        self.commands = dict()
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: Why not just {}

File: pajbot/managers/discord_bot.py [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+            self.redis.set("queued-subs-discord", json.dumps(data))
+
+    def add_command(self, *args, **kwargs):
+        cmd = Command(*args, **kwargs)
+        self.commands[cmd.name] = cmd
+        log.info('Added command %s', cmd)
+
+    async def _check(self, message):
+        for role in self.guild.get_member(message.author.id).roles:
+            if role.name == self.settings['admin_role']:
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: Definitely role IDs rather than names

File: pajbot/managers/discord_bot.py [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+                userconnections = db_session.query(UserConnections).filter_by(discord_user_id=str(message.author.id)).one()
+            except:
+                await self._private_message(message.author, 
+                    f"Your info has not been setup!"
+                )
+                return
+            user = User.find_by_id(db_session, userconnections.twitch_id)
+            if user.tier is None:
+                tier = 0
+            elif user.tier == '1000':
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: Can't you trim them into 1/2/3 before updating the user

File: pajbot/managers/discord_bot.py [hunk 11]
```diff
@@ -0,0 +1,363 @@
+import logging
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: Ask bulldog if he wants sub check on timer/when stream ends

File: pajbot/managers/discord_bot.py [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+        role_to_notify = None
+        role_to_ignore = None
+        for role in self.guild.roles:
+            if role.name == self.settings['twitchsubrole']:
+                sub_role = role
+            elif role.name == self.settings['tier2role']:
+                tier2_role = role
+            elif role.name == self.settings['tier3role']:
+                tier3_role = role
+            elif role.name == self.settings['role_to_notify']:
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: Some separate words with _ some don't separate, standardise it

File: pajbot/managers/discord_bot.py [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+            elif user.tier == '1000':
+                tier = 1
+            elif user.tier == '2000':
+                tier = 2
+            elif user.tier == '3000':
+                tier = 3
+            else:
+                tier = 0
+            await self.private_message(message.author, 
+                f'Tier {tier} sub:\nTwitch : {user} (<https://twitch.tv/{user.login}/>) \nDiscord : {message.author.display_name}#{message.author.discriminator} (<https://discordapp.com/users/{message.author.id}>)\nSteam : <https://steamcommunity.com/profiles/{userconnections.steam_id}/>'
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: I'm assuming here and below all those variables already exist and won't just display as None?

File: pajbot/managers/discord_bot.py [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+                    #notify role removal
+                    for member_to_notify in role_to_notify.members:
+                        user = User.find_by_id(db_session, unlinks['twitch_id'])
+                        steam_id = unlinks['steam_id']
+                        await self.private_message(member_to_notify,
+                            f'Account Data Unlinked : Tier 2 sub removal notification:\nTwitch : {user} (<https://twitch.tv/{user.login}/>)\nDiscord : {member.display_name}#{member.discriminator} (<https://discordapp.com/users/{member.id}>)\nSteam : <https://steamcommunity.com/profiles/{steam_id}/>'
+                        )
+            subs_to_return = []
+            for sub in queued_subs:
+                if datetime.strptime(sub[0], '%Y-%m-%d %H:%M:%S.%f') < datetime.now(): #must be run now
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: All datetime.now() should be utils.now()

File: pajbot/managers/discord_bot.py [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+                    member = self.guild.get_member(int(sub[1]))
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: Add comment somewhere describing contents of sub

File: pajbot/managers/discord_bot.py [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+                    else:
+                        if member_assigned_tier2 and tier2_role is not None:
+                            await self.remove_role(member, tier2_role)
+                        if member_assigned_tier3 and tier3_role is not None:
+                            await self.remove_role(member, tier3_role)
+                    if member_assigned_tier3 and quick_dict[member_id][0] != '3000' and self.settings['notify_on_unsub'] and self.settings['notify_on_tier3']:
+                        #notify role removal
+                        for member_to_notify in role_to_notify.members:
+                            await self.private_message(member_to_notify, 
+                                f'Tier 3 sub removal notification:\nTwitch : {User.find_by_id(db_session, quick_dict[member_id][1].twitch_id)} (<https://twitch.tv/{User.find_by_id(db_session, quick_dict[member_id][1].twitch_id).login}/>) \nDiscord : {member.display_name}#{member.discriminator} (<https://discordapp.com/users/{member.id}>)\nSteam : <https://steamcommunity.com/profiles/{quick_dict[member_id][1].steam_id}/>'
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: Tier {tier} sub removal would save a lot of hassle

File: pajbot/managers/discord_bot.py [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+                            if member_assigned_tier3:
+                                await self.remove_role(member, tier3_role)
+                            #notify role addition
+                            if self.settings['notify_on_new_sub'] and self.settings['notify_on_tier2']:
+                                for member_to_notify in role_to_notify.members:
+                                    await self.private_message(member_to_notify,
+                                        f'New tier 2 sub notification:\nTwitch : {User.find_by_id(db_session, quick_dict[member_id][1].twitch_id)}(<https://twitch.tv/{User.find_by_id(db_session, quick_dict[member_id][1].twitch_id).login}/>)\nDiscord : {member.display_name}#{member.discriminator} (<https://discordapp.com/users/{member.id}>)\nSteam : <https://steamcommunity.com/profiles/{quick_dict[member_id][1].steam_id}/>'
+                                    )
+                        if member_assigned_tier3 and quick_dict[member_id][0] != '3000':
+                            subs_to_return.append([(datetime.now() + timedelta(days=int(self.settings['grace_time']))).__str__(), member_id])
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: Christ

File: pajbot/migration_revisions/db/0012_discord_bot_modify [hunk 18]
```diff
@@ -0,0 +1,25 @@
+import logging
```
Comment (DatGuy1) [inline pajbot/migration_revisions/db/0012_discord_bot_modify]: .py

File: pajbot/models/user.py [hunk 19]
```diff
@@ -89,6 +90,8 @@ class User(Base):
     ignored = Column(BOOLEAN, nullable=False, server_default="FALSE")
     banned = Column(BOOLEAN, nullable=False, server_default="FALSE")
     timeout_end = Column(UtcDateTime(), nullable=True, server_default="NULL")
+    tier = Column(TEXT, nullable=True)
```
[the remark this answers is not in the document]
Comment (TroyDota) [reply pajbot/models/user.py]: well twitch returns it as "1000", "2000", "3000" so i didnt bother converting it to an int 1, 2, 3

Do you want me to do this?

File: pajbot/models/user.py (hunk 1 above)
Comment (TroyDota) [reply pajbot/models/user.py]: What do you mean?

File: pajbot/oauth_client_edit.py (hunk 4 above)
Comment (TroyDota) [reply pajbot/oauth_client_edit.py]: xd

File: uwsgi_shared.ini (hunk 6 above)
Comment (TroyDota) [reply uwsgi_shared.ini]: steam returns a bunch of shit and it failed without it. I dont know the exact value it should be i just googled the error and they told me to change it to this.

File: pajbot/models/user.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (DatGuy1) [reply pajbot/models/user.py]: Yes makes it more readable. Also postgres INTegers are 4 bytes

File: pajbot/managers/discord_bot.py (hunk 17 above)
Comment (TroyDota) [reply pajbot/managers/discord_bot.py]: xd

Comment (TroyDota) [reply pajbot/managers/discord_bot.py]: i changed it to utils.now()

File: pajbot/models/user.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (TroyDota) [reply pajbot/models/user.py]: k sadKEK

[the remark this answers is not in the document]
Comment (TroyDota) [reply pajbot/models/user.py]: Done hopefully

File: pajbot/managers/discord_bot.py (hunk 9 above)
Comment (TroyDota) [reply pajbot/managers/discord_bot.py]: You cant get role ids from the discord client

File: pajbot/managers/discord_bot.py (hunk 11 above)
Comment (TroyDota) [reply pajbot/managers/discord_bot.py]: Ill ask but it doesnt make much of a difference they get purged of roles every 10min if grace is none

File: pajbot/managers/discord_bot.py (hunk 13 above)
Comment (TroyDota) [reply pajbot/managers/discord_bot.py]: Yes.

File: pajbot/apiwrappers/twitch/helix.py (hunk 7 above)
Comment (TroyDota) [reply pajbot/apiwrappers/twitch/helix.py]: No?

File: pajbot/managers/discord_bot.py (hunk 11 above)
Comment (TroyDota) [reply pajbot/managers/discord_bot.py]: Oh wait
i read this wrong. I asked good suggestion

File: pajbot/managers/discord_bot.py (hunk 9 above)
Comment (TroyDota) [reply pajbot/managers/discord_bot.py]: So configuring the bot becomes tedious in the panel.

File: pajbot/apiwrappers/twitch/helix.py (hunk 7 above)
Comment (DatGuy1) [reply pajbot/apiwrappers/twitch/helix.py]: Run it then 4Head

**Comment by TroyDota** [comment]
I just thought that i should add a command for bot-admins to do !connections <twitch_username> and that will return the connection data for said user.

File: pajbot/managers/discord_bot.py (hunk 9 above)
Comment (DatGuy1) [reply pajbot/managers/discord_bot.py]: No? User -> roles -> right click -> Copy ID

Comment (TroyDota) [reply pajbot/managers/discord_bot.py]: I dont see it.

Comment (TroyDota) [reply pajbot/managers/discord_bot.py]: nevermind
ill change it

Comment (TroyDota) [reply pajbot/managers/discord_bot.py]: dev mode discord Pepega

File: pajbot/models/user.py (hunk 1 above)
Comment (TroyDota) [reply pajbot/models/user.py]: this is executed by the web panel. And adds a connection cooldown

File: pajbot/managers/discord_bot.py [hunk 20]
```diff
@@ -0,0 +1,356 @@
...
+        await self.bot.check_discord_roles()
+
+    async def on_message(self, message):
+        # If invite in private message, join server
+
+        if not message.content.startswith("!"):
+            return
+        data = message.content.split("!")
+        if len(data) <= 1:
+            log.debug('no command in message')
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: Comment these out

**Comment by TroyDota** [comment]
How do you verify the recent commits?

**Comment by TroyDota** [comment]
also if u want to do anything live. I work on stream and you can type live and ill update live. https://twitch.tv/troydota

File: pajbot/managers/discord_bot.py (hunk 8 above)
Comment (DatGuy1) [reply pajbot/managers/discord_bot.py]: Again?

**Comment by TroyDota** [comment]
Everything done

File: pajbot/managers/discord_bot.py [hunk 21]
```diff
@@ -0,0 +1,386 @@
...
+        await member.add_roles(role)
+
+    async def check_discord_roles(self):
+        try:
+            twitch_sub_role = None
+            tier2_role = None
+            tier3_role = None
+            notify_role = None
+            ignore_role = None
+            twitch_sub_role = self.guild.get_role(int(self.settings["twitch_sub_role"]))
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: If you use variables more than once you should declare them

File: pajbot/managers/discord_bot.py [hunk 22]
```diff
@@ -0,0 +1,386 @@
...
+
+    async def _remove_role(self, member, role):
+        await member.remove_roles(role)
+
+    async def _add_role(self, member, role):
+        await member.add_roles(role)
+
+    async def check_discord_roles(self):
+        try:
+            twitch_sub_role = None
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: Redefining?

File: pajbot/managers/discord_bot.py [hunk 23]
```diff
@@ -0,0 +1,386 @@
...
+                        member_assigned_tier3 = tier3_role in member.roles
+                    if member_assigned_tier3 and tier3_role is not None:
+                        await self.remove_role(member, tier3_role)
+                    if member_assigned_tier2 and tier2_role is not None:
+                        await self.remove_role(member, tier2_role)
+                    for member_to_notify in notify_role.members:
+                        user = User.find_by_id(db_session, unlinks["twitch_id"])
+                        steam_id = unlinks["steam_id"]
+                        message = (
+                            "Account Data Unlinked : Tier {tier} sub removal notification:\nTwitch : "
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: There's quite a few usages of concatenating strings and using f"" in same string

File: pajbot/managers/discord_bot.py [hunk 24]
```diff
@@ -0,0 +1,386 @@
...
+                            if twitch_sub_role not in member.roles or str(member.id) not in quick_dict:
+                                await self.remove_role(member, tier3_role)
+                data = {"array": subs_to_return}
+                self.redis.set("queued-subs-discord", json.dumps(data))
+                data = {"array": []}
+                self.redis.set("unlinks-subs-discord", json.dumps(data))
+                db_session.commit()
+            log.info("Discord roles Checked!")
+        except:
+            logging.exception("Discord Bot error")
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: log.exception

File: pajbot/modules/__init__.py (hunk 3 above)
Comment (DatGuy1) [reply pajbot/modules/__init__.py]: discord_bot doesn't go after wolfram

File: pajbot/web/routes/base/login.py [hunk 25]
```diff
@@ -158,11 +245,39 @@ def get_next_url(request, key="n"):
     def logout():
         session.pop("twitch_token", None)
         session.pop("user", None)
+        session.pop("steam_id", None)
+        session.pop("discord_token", None)
```
Comment (DatGuy1) [inline pajbot/web/routes/base/login.py]: Do you really need to store three discord variables?

File: pajbot/managers/discord_bot.py (hunk 24 above)
Comment (TroyDota) [reply pajbot/managers/discord_bot.py]: so we can see where it fails on first run without it hiding

File: pajbot/web/routes/base/login.py (hunk 25 above)
Comment (TroyDota) [reply pajbot/web/routes/base/login.py]: oauth requires it.

Comment (TroyDota) [reply pajbot/web/routes/base/login.py]: and the other 2 are for the connection and profile link on the connections page

**Comment by TroyDota** [comment]
The order of imports doesnt matter, and the imports are not in order to start with.

File: pajbot/managers/discord_bot.py (hunk 24 above)
Comment (TroyDota) [reply pajbot/managers/discord_bot.py]: we can remove this once the bot is 100%

File: pajbot/modules/__init__.py (hunk 3 above)
Comment (TroyDota) [reply pajbot/modules/__init__.py]: Imports arnt in order, but ill order all of them

File: pajbot/web/routes/base/login.py (hunk 25 above)
Comment (TroyDota) [reply pajbot/web/routes/base/login.py]: also i dont query the discord api ever again.

File: pajbot/managers/discord_bot.py (hunk 24 above)
Comment (DatGuy1) [reply pajbot/managers/discord_bot.py]: https://docs.python.org/3/library/logging.html#logging.Logger.exception

Comment (TroyDota) [reply pajbot/managers/discord_bot.py]: Okay well i changed it to log.error

**Comment by TroyDota** [comment]
I added back lines i removed that actually do something in other files

**Comment by TroyDota** [comment]
When do u think this will be merged and live?

**Comment by TroyDota** [comment]
I noticed a bug. Fixed before merge.

**Comment by TroyDota** [comment]
The last few fixes i applied havent been reviewed. Just bugs i noticed.

File: pajbot/managers/discord_bot.py [hunk 26]
```diff
@@ -49,31 +49,31 @@ def __init__(self, bot):
...
         cmd = self.bot.commands.get(data[1].split(" ")[0])
-
         if not cmd:
             return
-
-        # Go on.
-        await cmd.call(message)
+        try:
+            await cmd.call(message)
+        except Exception e:
```
Comment (DatGuy1) [inline pajbot/managers/discord_bot.py]: as e

**Pull request merged**